_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+This means attributes **cannot affect a rule's dependencies**.
+
+For example, C++ rules have an attribute named
+[`copts`](https://docs.bazel.build/versions/master/be/c-cpp.html#cc_binary.copts)
+that sets custom C++ compile options. `cc_binary(name = "mybinary", srcs =
+["mybinary.cc"], deps = [":mylibrary"], copts =
+["-DUSE_EXPERIMENTAL_FEATURES=1"])` might compile `mybinary.cc` with
+experimental features. But this won't happen for `:mylibrary` unless it sets
+its `copts` similarly. This makes attributes unsuitable for tasks like building
+a binary for a different CPU.
+
+If you imagine a build as a [graph](https://docs.bazel.build/versions/master/build-ref.html#dependencies)
+with parents above their dependencies, attributes change behavior
+[*up*](https://docs.bazel.build/versions/master/skylark/rules.html#providers)
+the graph while configuration changes behavior *down* the graph.
+
+## Platforms
+
+One of our team's major goals is designing a principled API for defining
+*platforms* and *toolchains*.
+
+While there are many kinds of "ways" you might want to build your project, in
+practice most developers want the flexibility to target different devices,
+OSes, CPUs, and other machine properties. The general term for this is
+*platforms*. Being able to reason about how your code interacts with platforms
+is the basis of [multi-platform software](https://en.wikipedia.org/wiki/Cross-platform_software).
+
+*Toolchains* are the set of programs that build your code. A C++ toolchain
+includes a compiler and linker and the flags that invoke them "correctly".
+Different platforms require different toolchains. gcc might be the compiler of
+choice for Linux while Xcode is used on the Mac. Even the same platform might
+use different toolchains. For example, maybe you want to use an experimental
+compiler.
+
+Not only are these concepts cross-language, but it's *important* that languages
+treat them consistently. Otherwise you get an uncoordinated mess of
+language-specific code that makes it hard to combine different languages in a
+single project.
+
+One of the reasons
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+doesn't work well is that `--cpu` and `--crosstool_top` are C++-specific. What
+if you want to build Java libraries with C++ dependencies for both Linux and
+Android? This impacts not just C++, but also the JDK and maybe even which
+support libraries you need.
+
+Not only are `--cpu` and `-crosstool_top` insufficiently expressive, but they
+might actively sabotoge you. Consider the following use of
+[select](https://docs.bazel.build/versions/master/configurable-attributes.html):
+
+```python
+java_library(
+    name = "supporting_library",
+	deps = select({
+	    ":android": [":extra_required_android_deps"],
+		"//conditions:default": []
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
```

Comment:
Since ":android" isn't defined in this example, it's not clear how this differs from the modification below. I think this difference is critical to motivating the history explanation here and the payoff below.

Reply:
Done.